src/js/abstract/hocs/with-context.js

```diff
@@ -0,0 +1,113 @@
+import { publish, subscribe } from '../../pubsub';
+import lifecycleLogger from '../utils/lifecycle-logger';
+
+const withContext = Base =>
+  /**
+   * Adds the ability to provide and consume contextual data.
+   */
+  class Context extends Base {
+    /**
+     * connectedCallback - description
+     *
+     * @return {type}  description
+     */
+    connectedCallback() {
+      if (super.connectedCallback) {
+        super.connectedCallback();
+      }
+
+      if (this.contextCallback) {
+        this._makeContextReady();
+      }
+    }
+
+    /**
+     * disconnectedCallback - description
+     *
+     * @return {type}  description
+     */
+    disconnectedCallback() {
+      if (super.disconnectedCallback) {
+        super.disconnectedCallback();
+      }
+
+      if (this.unContextEnabled) {
+        this.unContextEnabled();
+      }
+    }
+
+    // @TODO: atm no data can be shared by enabling context, though this could be necessary
+    /**
+     * Provides an opt-in contextual scope for hierarchy-agnostic child components.
+     */
+    enableContext() {
+      if (ENV !== PROD) {
+        lifecycleLogger(this.logLifecycle)(`enableContext -> ${this.nodeName}#${this._id}`);
+      }
+
+      const contextName = this.nodeName.toLowerCase();
+
+      this.__isContext = true;
+      this.__contextName = contextName;
+
+      // publish context/enabled with contextual node name
+      publish('context/enabled', contextName);
+    }
+
+    /**
+     * Opt-in to select a specific context by component name.
+     *
+     * @param {String} name - `nodeName` of the desired contextual node.
+     */
+    selectContext(name) {
```

Comment:
@LucaMele 
@TheDadi 
Same for `selectContext`, may change to `consumeContext`?

Reply:
naming is better like that yes